### PR TITLE
Impl: reduce syscall and memcopy for multiple package

### PR DESCRIPTION
**Pull request by watermelo**
origin issue https://github.com/AlexStocks/getty/issues/33

### CODE REVIEW

**Comment by gaoxinge** [comment]
LGTM

**Comment by wongoo** [comment]
`net.Buffers()` is a great implement

File: session.go [hunk 1]
```diff
@@ -449,6 +449,13 @@ func (s *session) WriteBytesArray(pkgs ...[]byte) error {
 		return s.WriteBytes(pkgs[0])
 	}
 
+	// reduce syscall and memcopy for multiple packages
+	if _, ok := s.Connection.(*gettyTCPConn); ok {
+		if _, err := s.Connection.send(pkgs); err != nil {
+			return perrors.Wrapf(err, "s.Connection.Write(pkgs num:%d)", len(pkgs))
+		}
+	}
+
```
Comment (wongoo) [inline session.go:458]: `WriteBytesArray()` is only supported by tcp connection,  the code below is not needed.

Comment (watermelo) [reply session.go:458]: > `WriteBytesArray()` is only supported by tcp connection, the code below is not needed.

Looks like websocket also used it.

File: connection.go [hunk 2]
```diff
@@ -281,11 +278,27 @@ func (t *gettyTCPConn) send(pkg interface{}) (int, error) {
...
+		if length, err = t.writer.Write(p); err == nil {
+			atomic.AddUint32(&t.writeBytes, (uint32)(len(p)))
+		}
+		log.Debug("localAddr: %s, remoteAddr:%s, now:%s, length:%d, err:%s",
+			t.conn.LocalAddr(), t.conn.RemoteAddr(), currentTime, length, err)
+		return length, perrors.WithStack(err)
+	}
+
+	return 0, perrors.Errorf("illegal @pkg{%#v} type", pkg)
 	//return length, err
```
Comment (fangyincheng) [inline connection.go:302]: delete unuseful  line ？

**Comment by gaoxinge** [comment]
@watermelo Can `WriteBytes` and `WriteBytesArray` in `session.go` be unified to one function? They look very similar.

**Comment by gaoxinge** [comment]
@AlexStocks @watermelo  嗯，是的。还有一个小小的问题，就是`session.go`里面的`incWritePkgNum()`

https://github.com/dubbogo/getty/blob/8b02420a4bc8f7928b22ccf6fb1caaecab5b2e8e/session.go#L431-L439

需要放到下面的用`connection.go`里面

https://github.com/dubbogo/getty/blob/8b02420a4bc8f7928b22ccf6fb1caaecab5b2e8e/connection.go#L292-L299

这样看起来是不是统一一点？

**Comment by watermelo** [comment]
> @AlexStocks @watermelo 嗯，是的。还有一个小小的问题，就是`session.go`里面的`incWritePkgNum()`
> 
> https://github.com/dubbogo/getty/blob/8b02420a4bc8f7928b22ccf6fb1caaecab5b2e8e/session.go#L431-L439
> 
> 是不是需要放到下面的用`connection.go`里面？
> 
> https://github.com/dubbogo/getty/blob/8b02420a4bc8f7928b22ccf6fb1caaecab5b2e8e/connection.go#L292-L299
> 
> 这样看起来是不是统一点？

@gaoxinge done

**Comment by gaoxinge** [comment]
@watermelo Great job! LGTM.

**Pull request merged**